File: libplugin/src/main/java/com/lemon/plugin/inject/tool/javaassist/JavaAssistInjectTool.java
package com.lemon.plugin.inject.tool.javaassist;

import com.lemon.plugin.inject.InjectUtils;
import com.lemon.plugin.inject.IInjectTool;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Method;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.HashSet;

import javassist.CannotCompileException;
import javassist.ClassPool;
import javassist.CtClass;
import javassist.NotFoundException;

public class JavaAssistInjectTool implements IInjectTool {
    private ClassPool mClassPool;
    private final HashSet<String> mModuleSet = new HashSet<>();

    private URLClassLoader mClassLoader;
    private final Method mURLMethod;

    private final String mRootPath;
    private final Injector mInjector = new Injector();

    private String mLastFlavor = "";

    public JavaAssistInjectTool(String rootPath) {
        mRootPath = rootPath;

        mModuleSet.add("libjava");
        mModuleSet.add("libandroid");

        try {
            mURLMethod = URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
            mURLMethod.setAccessible(true);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void injectDir(File dir) {
        initAssistFlavor(dir);

        initAssistDir(dir);
        initClassLoaderDir(dir);

    }

    private void initAssistFlavor(File dir) {
        String dirPath = dir.getAbsolutePath();
        int index = dirPath.lastIndexOf('/');
        String flavor = dirPath.substring(index + 1, dirPath.length());

        if (flavor.equals(mLastFlavor)) {
            return;
        }

        mClassPool = new ClassPool(null);
        mClassPool.appendSystemPath();

        mClassLoader = new URLClassLoader(new URL[0], getClass().getClassLoader());
        mInjector.setClassLoader(mClassLoader);

        for (String module : mModuleSet) {
            String path = mRootPath + "/" + module + "/build/intermediates/classes/" + flavor;
            InjectUtils.println("lib path:" + path);
            try {
                mClassPool.appendClassPath(path);
            } catch (NotFoundException e) {
                throw new RuntimeException(e);
            }
        }

        mLastFlavor = flavor;
    }

    private void initClassLoaderDir(File dir) {
        try {
            mURLMethod.invoke(mClassLoader, dir.toURI().toURL());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void initAssistDir(File dir) {
        try {
            mClassPool.appendClassPath(dir.getAbsolutePath());
        } catch (NotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void injectFile(File dir, File file) {
        String dirPath = dir.getAbsolutePath();
        String filePath = file.getAbsolutePath();
        String className = getClassName(dirPath, filePath);

        try {
            CtClass clazz = mClassPool.getCtClass(className);
            if (clazz.isFrozen()) {
                clazz.defrost();
            }
            boolean inject = mInjector.inject(filePath, clazz);
            if (inject) {
                clazz.writeFile(dirPath);
            }
        } catch (NotFoundException | CannotCompileException | IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static String getClassName(String dirPath, String filePath) {
        filePath = filePath.substring(dirPath.length() + 1);
        filePath = filePath.replace('/', '.').replace('\\', '.');
        filePath = filePath.substring(0, filePath.length() - ".class".length());
        return filePath;
    }
}

File: libplugin/src/main/groovy/com/creation/plugin/inject/tool/javaassist/JavaAssistInjectTool.kt
package com.creation.plugin.inject.tool.javaassist

import com.creation.plugin.inject.InjectUtils
import com.creation.plugin.inject.IInjectTool

import java.io.File
import java.io.IOException
import java.lang.reflect.Method
import java.net.URL
import java.net.URLClassLoader
import java.util.HashSet

import javassist.CannotCompileException
import javassist.ClassPool
import javassist.NotFoundException

class JavaAssistInjectTool(private val mRootPath: String) : IInjectTool {
    private lateinit var mClassPool: ClassPool
    private val mModuleSet = HashSet<String>()

    private lateinit var mClassLoader: URLClassLoader
    private val mURLMethod: Method
    private val mInjector = Injector()

    private var mLastFlavor = ""

    init {

        // add
//        mModuleSet.add(null)

        try {
            mURLMethod = URLClassLoader::class.java.getDeclaredMethod("addURL", URL::class.java)
            mURLMethod.isAccessible = true
        } catch (e: NoSuchMethodException) {
            throw RuntimeException(e)
        }

    }

    override fun injectDir(dir: File) {
        initAssistFlavor(dir)

        initAssistDir(dir)
        initClassLoaderDir(dir)

    }

    private fun initAssistFlavor(dir: File) {
        val dirPath = dir.absolutePath
        val index = dirPath.lastIndexOf('/')
        val flavor = dirPath.substring(index + 1, dirPath.length)

        if (flavor == mLastFlavor) {
            return
        }

        mClassPool = ClassPool(null)
        mClassPool.appendSystemPath()

        mClassLoader = URLClassLoader(arrayOfNulls(0), javaClass.classLoader)
        mInjector.setClassLoader(mClassLoader)

        for (module in mModuleSet) {
            val path = "$mRootPath/$module/build/intermediates/classes/$flavor"
            InjectUtils.println("lib path:$path")
            try {
                mClassPool.appendClassPath(path)
            } catch (e: NotFoundException) {
                throw RuntimeException(e)
            }

        }

        mLastFlavor = flavor
    }

    private fun initClassLoaderDir(dir: File) {
        try {
            mURLMethod.invoke(mClassLoader, dir.toURI().toURL())
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun initAssistDir(dir: File) {
        try {
            mClassPool.appendClassPath(dir.absolutePath)
        } catch (e: NotFoundException) {
            throw RuntimeException(e)
        }

    }

    override fun injectFile(dir: File, file: File) {
        val dirPath = dir.absolutePath
        val filePath = file.absolutePath
        val className = getClassName(dirPath, filePath)

        try {
            val clazz = mClassPool.getCtClass(className)
            if (clazz.isFrozen) {
                clazz.defrost()
            }
            val inject = mInjector.inject(filePath, clazz)
            if (inject) {
                clazz.writeFile(dirPath)
            }
        } catch (e: NotFoundException) {
            throw RuntimeException(e)
        } catch (e: CannotCompileException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }

    private fun getClassName(dirPath: String, path: String): String {
        var filePath = path.substring(dirPath.length + 1)
        filePath = filePath.replace('/', '.').replace('\\', '.')
        filePath = filePath.substring(0, filePath.length - ".class".length)
        return filePath
    }
}
